libs/core/async_sycl/include/hpx/async_sycl/sycl_executor.hpp

```diff
@@ -0,0 +1,182 @@
+
+#include <cstddef>
+#include <exception>
+#include <memory>
+#include <type_traits>
+#include <utility>
+
+namespace hpx { namespace sycl { namespace experimental {
+
+    struct sycl_executor
+    {
+        using future_type = hpx::future<void>;
+
+        // -------------------------------------------------------------------------
+        /// Create a SYCL executor (based on a sycl queue)
+        explicit sycl_executor(cl::sycl::default_selector selector)
+          : command_queue(selector, cl::sycl::property::queue::in_order{})
+        {
+        }
+
+        // TODO Future work: Add more constructors as needed (in case we need different queue constructors)
+        // -------------------------------------------------------------------------
+        // Queue will be cleaned up by its own destructor
+        ~sycl_executor() {}
+
+        /// Get future for this command_queue (NOTE will be more efficient if
+        /// an event is provided -- otherwise a dummy kernel must be submitted to
+        /// get an event)
+        future_type get_future()
+        {
+            // The SYCL standard does not include a eventRecord method Instead
+            // we have to submit some dummy function and use the event the
+            // launch returns
+            cl::sycl::event event = command_queue.submit(
+                [&](cl::sycl::handler& h) { h.single_task([]() {}); });
+            return detail::get_future(event);
+        }
+
+        /// Get future for that becomes ready when the given event completes
+        inline future_type get_future(cl::sycl::event event)
+        {
+            return detail::get_future(event);
+        }
+
+#if defined(__INTEL_LLVM_COMPILER)
+        // To find the correct overload (or any at all actually) we need to add the
+        // code_location argument which is the last argument in every queue member
+        // function in the intel oneapi sycl implementation.  As far as I can tell
+        // it is usually invisible from the user-side since it is using a default
+        // argument (code_location::current())
+
+        /// sycl::queue::member_function type with code_location parameter
+        template <typename... Params>
+        using queue_function_ptr_t = cl::sycl::event (cl::sycl::queue::*)(
+            Params..., const cl::sycl::detail::code_location&);
+        /// Invoke member function given queue and parameters. Default
+        /// code_location argument added automatically.
+        template <typename... Params>
+        void post(queue_function_ptr_t<Params...>&& queue_member_function,
+            Params&&... args)
+        {
+            // for the intel version we need to actually pass the code location.
+            // Within the intel sycl api this is usually a default argument,
+            // but for invoke we need to pass it manually 
+            cl::sycl::event e =
+                std::invoke(std::forward<queue_function_ptr_t<Params...>>(
```

Comment:
Would you mind using `HPX_INVOKE()` instead of `std::invoke()` (both are functionally fully compatible)? Our implementation is much less heavyweight in terms of compilation overheads. (please note, that we use `HPX_INVOKE` only inside the library, not for examples or tests).

Reply:
I have done so in e280fd14c3ab236ed640c3b58dac7236308c821d

I also changed std::forward to HPX_FORWARD while I was at it! Is the reasoning behind HPX_FORWARD the same? (better compiletime)

Reply:
Yes, it's just all about compile time overheads. Thanks!